Sources/SourceKitLSP/Swift/DiagnosticReportManager.swift

```diff
@@ -0,0 +1,175 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LSPLogging
+import LanguageServerProtocol
+import SourceKitD
+import SwiftParserDiagnostics
+
+actor DiagnosticReportManager {
+  private let sourcekitd: SourceKitD
+  private let syntaxTreeManager: SyntaxTreeManager
+  private let documentManager: DocumentManager
+  private let clientHasDiagnosticsCodeDescriptionSupport: Bool
+
+  private nonisolated var keys: sourcekitd_keys { return sourcekitd.keys }
+  private nonisolated var requests: sourcekitd_requests { return sourcekitd.requests }
+
+  /// The cache that stores reports for snapshot id and buildSettings
+  ///
+  /// Conceptually, this is a dictionary. To prevent excessive memory usage we
+  /// only keep `cacheSize` entries within the array. Older entries are at the
+  /// end of the list, newer entries at the front.
+  private var reportCache:
+    [(
+      snapshotID: DocumentSnapshot.ID,
+      buildSettings: SwiftCompileCommand?,
+      report: RelatedFullDocumentDiagnosticReport
+    )] = []
+
+  /// The number of reports to keep
+  ///
+  /// - Note: This has been chosen without scientific measurements.
+  private let cacheSize = 5
+
+  init(
+    sourcekitd: SourceKitD,
+    syntaxTreeManager: SyntaxTreeManager,
+    documentManager: DocumentManager,
+    clientHasDiagnosticsCodeDescriptionSupport: Bool
+  ) {
+    self.sourcekitd = sourcekitd
+    self.syntaxTreeManager = syntaxTreeManager
+    self.documentManager = documentManager
+    self.clientHasDiagnosticsCodeDescriptionSupport = clientHasDiagnosticsCodeDescriptionSupport
+  }
+
+  func diagnosticReport(for snapshot: DocumentSnapshot, buildSettings: SwiftCompileCommand?, useCache: Bool = false)
+    async throws -> RelatedFullDocumentDiagnosticReport
+  {
+    if useCache, let report = report(for: snapshot.id, buildSettings: buildSettings) {
+      return report
+    }
+    guard let buildSettings = buildSettings else {
+      logger.log(
+        "Producing syntactic diagnostics from the built-in swift-syntax because we don't have buildSettings"
+      )
+      // If we don't have build settings,
+      // sourcekitd won't be able to give us accurate semantic diagnostics.
+      // Fall back to providing syntactic diagnostics from the built-in
+      // swift-syntax. That's the best we can do for now.
+      let report = try await requestFallbackReport(with: snapshot)
```

Comment:
There’s currently a (kind of theoretical) race condition and duplicated computation here: If we request two diagnostic reports for the same file and build settings, both of them trigger `requestFallbackReport` and then both of them call `setReport`. While `requestFallbackReport` should be returning the same data for both calls, if it didn’t, we would get inconsistent results for the two calls.

To avoid the duplicated computation and the race condition here, `SyntaxTreeManager` has the `syntaxTreeComputations` dictionary. Because we map each `snapshotID` to a `Task`, we can insert a value to it immediately when the syntax tree is requested and thus we can guarantee that every syntax tree is only every built once. I think we should do the same here (in particular because getting diagnostics from sourcekitd can be a lot more expensive the building a syntax tree).

The same applies to the `requestReport` call below.

Reply:
i thought the cache in an actor would be thread-safe, but i can see we're using Task in SyntaxTreeManager.
let me follow the way we're doing first.

Reply:
Actors are thread-safe for low-level data races (where multiple threads try to access the same variable simultaneously) but they can still have high-level data races. These occur if you store some kind of state in local variables but that state can change at suspension points, aka. `await` calls. That’s why you explicitly spell out `await` in the call.